script/testing/README.md

```diff
@@ -54,15 +55,75 @@ If you specify the `--query-mode extended`, you then can also indicate the prepa
 - run the post-suite task (test suite specific) 
 - print out the logs to the stdout
 
-### Adding a new test case
-All test cases should inherit from the `TestCase` class. Anyone is free to modify any attribute from the base class.
-- Mandatory attributes
-  - `test_command` (`List(str)`): the command to run the test case
-- Optional attributes
-  - `test_command_cwd` (`str`): the working directory to run the test command
-  - `test_error_msg` (`str`): the error message to display in case of errors
-- Optional functions
-  - `run_pre_test`: the pre-test tasks required for the test
-    - config the xml file, etc.
-  - `run_post_test`: the post-test tasks required for the test
-    - e.g. parse the output json, etc.
+## How to create a new test type
+The classes in the `util` folder can be used and extend to help you create a new test type.
+
+### Base classes
+- `NoisePageServer`
+  - Manage the lifecycle of the NoisePage instance. It create a Python subprocess for the NoisePage process, poll the logs, and terminate or kill it when the test finishes
+- `TestCase`
+  - Manage a the life cycle of a test case. A test case is usually a process trigger from command line. The actual test case can be specified as a command string and created and executed in a Python subprocess. 
+  - The `TestCase` class also provides `run_pre_test` and `run_post_test` functions for you to override for preparation and clean up of each test case.
+- `TestServer`
+  - Manage the entire lifecycle of a test. It uses the `NoisePageServer` to manage the database process. One `TestServer` can have a list of `TestCase`s and treats the entire collection as a suite. 
+  - It also provides the `run_pre_suite` and `run_post_suite` functions for you to override to specify any preparation and cleanup at the suite level.
+
+### Step-by-step instructions
+- Create the folder for your test under `noisepage/script/testing/<mytest>`
+- In the folder of your test, create the following files
+  - `run_<mytest>.py`
+    - The entry script for your test, which specifies the command line arguments and options.
+    - You can refer to [oltpbench/run_oltpbench.py](https://github.com/cmu-db/noisepage/blob/master/script/testing/oltpbench/run_oltpbench.py) for reference.
+  - `test_<mytest>.py`
```

Comment:
You should mention that this is only necessary if extending TestServer. right?

Reply:
Not really. `TestCase` is no longer optional once we introduced `TestSuite` with multiple `TestCase` structure.